pkg/cover/backend/elf.go

```diff
@@ -417,23 +424,65 @@ func readCoverPoints(file *elf.File, info *symbolInfo) ([2][]uint64, error) {
 	if err != nil {
 		return pcs, fmt.Errorf("failed to read .text: %v", err)
 	}
-	end := len(data) - callLen + 1
+	end := len(data)
+	if arch == targets.AMD64 {
+		end = end - callLen + 1
+	}
 	for i := 0; i < end; i++ {
-		pos := bytes.IndexByte(data[i:end], 0xe8)
+		pos := bytes.IndexByte(data[i:end], opcode)
 		if pos == -1 {
 			break
 		}
 		pos += i
 		i = pos
-		off := uint64(int64(int32(binary.LittleEndian.Uint32(data[pos+1:]))))
 		pc := text.Addr + uint64(pos)
-		target := pc + off + callLen
+		var off uint64
+		var target uint64
+		if arch == targets.AMD64 {
+			off = uint64(int64(int32(binary.LittleEndian.Uint32(data[pos+1:]))))
+			target = pc + off + uint64(callLen)
+		} else if arch == targets.ARM64 {
+			b := data[pos-callLen+1 : pos]
```

Comment:
This can crash if `0x94` is located in the first bytes. I think we need:
```
if pos < callLen-1 {
  continue
}
```
But also this is the case we avoid with `end = end - callLen + 1` for amd64. So now we have both opcode in the beginning and at the end of instruction, it would be more consistent to do:
```
for i := 0; i < len(data); i++ {
  pos := bytes.IndexByte(data[i:], opcode)
  ...
  if arch == targets.AMD64 {
    if pos > len(data) - callLen {
      continue
    }
    ....
  } else if arch == targets.ARM64 {
    if pos < callLen-1 {
      continue
    }
    ...
  }
}
```

or, perhaps, taking into account the next comment re pc adjustment, it looks even better to do:
```
for i := 0; i < len(data); i++ {
  pos := bytes.IndexByte(data[i:], opcode)
  ...
  if arch == targets.ARM64 {
    // opcode is at the end of instruction.
    pos -= callLen - 1
    if pos < 0 {
      continue
    }
  }
  if pos > len(data) - callLen {
    continue
  }
  pc := text.Addr + uint64(pos)
  ...
}
```
Then we have all proper pos checks in place and don't need to adjust pos/pc 3 times for arm64.